### PR TITLE
Fixed numeric parameter substitution bug  - Fixes #317 and #349

### LINKED ISSUE #317: Parameter substitution issue
foo = "one" 
bar = "this is a string :1"
cursor.execute("insert into table 1 values(%s, %s)", [foo, bar])

:1 in string gets substituted with "one".

### PR DESCRIPTION
**Pull request by MarechJ**
1) For direct use of `numbe`r paramstyle fixed by making the substitution operation from a loop to a one shot regex substitution.
2) For qmark or format paramstyle: After the substition already happens successfully but gets processed again at the numeric paramstyle step. 
Fixed (workaround) by allowing the user to explicitly define the paramstyle he is using.
Bug 2) could be improved by sniffing the paramstyle or sticking with the default one and just stick with this one instead of trying to solve everything and allowing a mix of several styles?

Please advise for running your tests. Do you use a local instance running in docker?

Usage of solution 2):

`cursor.execute("INSERT INTO test_impyla VALUES (?, ?)", ['this will not go wrong:2', 'im not going to creep in'], configuration={'paramstyle': 'qmark'})`

For 1) it's seamless

 - Fixes #317 and #349

### CODE REVIEW

**Comment by timarmstrong** [comment]
Can you please include tests for the bugs you fixed?

**Comment by MarechJ** [comment]
I need guidance for testing. I don't have a dev environment for impala. Do you run a local instance using the minimal image there? https://hub.docker.com/r/cloudera/impala-dev

**Comment by timarmstrong** [comment]
I missed your reply somehow. I think you basically need an impala cluster to point the tests at  - you configure via env variables then run the tests. But a dev environment might be the easiest way to achieve that. I think that docker image is stale- the canonical reference for setting up an impala dev environment is the Apache impala wiki - https://cwiki.apache.org/confluence/display/IMPALA/Impala+Home

I realise that's a PITA potentially to do just for this change so happy to work with you - e.g. even if you provide me the queries you used to manually test this, I could add them to the test framework and run the tests.

**Comment by MarechJ** [comment]
Hi, No worries. 

Indeed that'd save me quite some time not to install a dev env. Otherwise I'll have to do that after other higher prio things are done.

The query I use to test is: 

```
CREATE TABLE test_impyla
 ( acol                     string,  
   anothercol                     string )
STORED AS parquet
```
```
cursor.execute("INSERT INTO test_impyla VALUES (?, ?)", ['this will go wrong:2', 'im going to creep in'])
cursor.execute("INSERT INTO test_impyla VALUES (:1, :2)", ['this will go wrong:1', 'im going to creep in'])
```
Before the patch those would just crash.

And for checking the new option:
```
cursor.execute("INSERT INTO test_impyla VALUES (?, ?)", ['this will go wrong:1', 'im going to creep in'], configuration={'paramstyle': 'qmark'})
```

Is that enough?

**Comment by timarmstrong** [comment]
Sorry for the delay, things have been hectic here. 

It actually turns out that you can run the parameter substitution tests without an Impala instance by running "py.test impala/tests/test_query_parameters.py" - they just check the string produced by substitution. I added a few tests to get started here - https://github.com/timarmstrong/impyla/tree/MarechJ-master. Could you incorporate those tests into your PR and maybe include a few more, e.g. other cases where there is still a bug.

**Comment by MarechJ** [comment]
Yes indeed, I was too focused on integration test. I'll do that in the coming days.

File: impala/tests/test_query_parameters.py [hunk 1]
```diff
@@ -240,6 +240,44 @@ def test_format():
...
+     ['string :2 ? %s :named', '42'], paramstyle='qmark')
+
+  dt("select ?, %s, :named from test where a='string' and b='42'",
+     "select ?, %s, :named from test where a=:1 and b=:2",
+     ['string', '42'], paramstyle='numeric')
+  dt("select * from test where a='string :2 ? %s :named' and b='42'",
+      "select * from test where a=:1 and b=:2",
+     ['string :2 ? %s :named', '42'], paramstyle='numeric')
+
+  # BUG: %s picks up named parameters as stringified dict when dict is passed.
```
[the remark this answers is not in the document]
Comment (timarmstrong) [reply impala/tests/test_query_parameters.py]: It's a weird edge case - I don't think it makes a lot of sense to substitute %s when the arguments are a dict, but this seems to be what happens. Anyway, I think it makes sense to leave this test case in to document the current behaviour.

**Comment by timarmstrong** [comment]
Thanks for the contribution. I think this helps a lot in providing a workaround and some more options to control parameter substitution. I think I'd like to leave one of the issues open since the double substitution is still an issue (we should really be doing a single pass substitution).

File: impala/tests/test_query_parameters.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarechJ) [reply impala/tests/test_query_parameters.py]: ok

**Comment by MarechJ** [comment]
> Thanks for the contribution. I think this helps a lot in providing a workaround and some more options to control parameter substitution. I think I'd like to leave one of the issues open since the double substitution is still an issue (we should really be doing a single pass substitution).

You're welcome. Thanks for the tests. What is the schedule for releases?

**Comment by timarmstrong** [comment]
We haven't really had a regular release cadence. 0.15.0 was done fairly recently so we'll probably want to accumulate a few more changes to really justify a release.
